Game Framework

Manuzor opened the pull request:
The idea is to have a proper game framework where we can create new games fast in a flexible way.

Oh and... whoops, did some more work than just the game framework. :stuck_out_tongue:

Manuzor committed:
Add GameLoop class.

Also util to get the current working dir.

Manuzor committed:
Implement basics of GlobalGameLoopRegistry.

No dependency management yet.

Manuzor committed:
Define main module structure

Default main module is not fully implemented yet though.

Manuzor committed:
Use ezSubSystem for main module

This allows for greater flexibility since the main module is now a sub system and can be integrated in the existing dependency management chain.

Manuzor committed:
HTML Log

Fix kr::cwd off-by-one error.
Mount current working dir as category ".".
Write html log on core startup. Might be too late then?

pampersrocker commented:
Currently crashes on release: 
```
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /IC:\workspace\krepel\code\krEngine\.. /I"C:\workspace\toolbox\glew-1.12.0\include" /I"C:\workspace\toolbox\ezEngine-rev1196\Code\Engine" /nologo /W3 /WX /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D "CMAKE_INTDIR=\"Release\"" /D KR_ENGINE_DLL_EXPORT /D _WINDLL /D _MBCS /Gm- /EHsc /MD /GS /fp:fast /Zc:wchar_t /Zc:forScope /Zc:inline /GR- /Yc"C:/workspace/krepel/code/krEngine/pch.h" /Fp"krEngine.dir\Release\krEngine.pch" /Fo"krEngine.dir\Release\\" /Fd"krEngine.dir\Release\vc140.pdb" /Gd /TP /FIC:/workspace/krepel/code/krEngine/pch.h /errorReport:queue  \"/INCREMENTAL:NO\" \"/OPT:REF\" \"/OPT:ICF\" C:\workspace\krepel\workspace\code\krEngine\pch.cpp
  INCREMENTAL:NO"
c1xx : fatal error C1083: Cannot open source file: '"/INCREMENTAL:NO"': No such file or directory [C:\workspace\krepel\workspace\code\krEngine\krEngine.vcxproj]
  OPT:REF"
c1xx : fatal error C1083: Cannot open source file: '"/OPT:REF"': No such file or directory [C:\workspace\krepel\workspace\code\krEngine\krEngine.vcxproj]
  OPT:ICF"
c1xx : fatal error C1083: Cannot open source file: '"/OPT:ICF"': No such file or directory [C:\workspace\krepel\workspace\code\krEngine\krEngine.vcxproj]
  pch.cpp
```

Manuzor committed:
Fix release builds

Setting of linker options was done incorrectly and CMake proves to be an inconsistent son of a bee once more...

Manuzor committed:
Easier custom data directories

Provide helper function to get the default root path as an absolute native path.

Manuzor committed:
Util makePath

To prevent erroneous use of ezStringBuilder to create paths.

Manuzor committed:
Include the launcher exe on krepel user-side

Manuzor committed:
Add kr_target_make_launchable

Manuzor committed:
kr::configPostfix() + usage in krGameLauncher

The function returns -debug, etc. as a string, which is used to load the proper game modules in the current config mode, e.g. asteroids-debug.

Manuzor committed:
Generalize pullRuntimeBinaries.

pullRuntimeBinaries has been generalized to work with krepel and outside project as well. kr_client_copy_dlls for krepel projects is now obsolete. This makes it easier for client projects to retrieve runtime files whenever they want.

Manuzor committed:
DefaultMainModule now uses kr::Window.

Instead of using ezWindow directly.

Manuzor committed:
Make use of ezClock.

The state in this commit is not to be taken to seriously, as things will change a lot soon.

Manuzor committed:
GlobalGameLoopRegistry takes callbacks directly now.

kr::GameLoop was removed and the global registry now associates a name with a given callback.
Dependency management is still NOT implemented yet.
Also fix issue with global clocks. Renderer and module now keep their own instances.

Manuzor committed:
Fix log message formatting.

Manuzor committed:
Game loop priority.

Game loops can now be sorted by priority, which ist just a signed 32 bit integer.
I refrained from implemented a full-blown dependency management system as it is probably overkill for us.
There is a convenience function to print the tick order, which can be used for debugging (kr::GlobalGameLoopRegistry::printTickOrder).

Manuzor committed:
Rename kr::GlobalGameLoop{ Registry => }

Because by now its not just a registry anymore.

Manuzor committed:
Make GlobalGameLoop a namespace instead of a struct.

Also add some docs for kr::DefaultMainModule::tick.

Manuzor committed:
Proper Project Template

We now have a proper project template (as opposed to having everything hardcoded as string within a single python script). Every files' name and the contents of the files that belong to the project template can use the substitution syntax ~{var_name}.

The krGameLauncher is basically obsolete now because it is always generted to the new projects and will be called like the project with a `_launcher` suffix.

Manuzor committed:
Remove krGameLauncher

as well as GameTest1 which depended on it.

Manuzor committed:
Remove last remnants of krGameLauncher.

Manuzor committed:
Add default 2D camera.

Manuzor commented:
I think it's time to take this code right into some action to see how we can extend the framework. There are still several parts missing like easier rendering interface, but that also depends on our rendering code, which will be heavily refactored soon I believe.

I want to test the current framework with asteroids for example to see how to improve it.
